File: ui/src/api/process/process.ts
import request from '@/utils/request'

const BASE_URL = '/api/v1/process'

const ProcessAPI = {
  /** 获取数据 */
  getList(queryParams?: ProcessQuery) {
    return request({
      url: `${BASE_URL}`,
      method: 'get',
      params: queryParams,
    })
  },
  getDetail(id: number) {
    return request({
      url: `${BASE_URL}/${id}`,
      method: 'get',
    })
  },
  /** 添加 */
  create(data: ProcessForm) {
    return request({
      url: `${BASE_URL}`,
      method: 'post',
      data: data,
    })
  },

  /**
   * 更新
   *
   * @param id ID
   * @param data 表单数据
   */
  update(data: ProcessForm) {
    return request({
      url: `${BASE_URL}`,
      method: 'put',
      data: data,
    })
  },

  /**
   * 删除
   *
   * @param id ID
   */
  delete(id: number) {
    return request({
      url: `${BASE_URL}/${id}`,
      method: 'delete',
    })
  },
  copy(id: number) {
    return request({
      url: `${BASE_URL}/copy/${id}`,
      method: 'post',
    })
  },
}

export default ProcessAPI

/** 查询参数 */
export interface ProcessQuery {
  /** 搜索关键字 */
  page?: number
  pageSize?: number
  status?: number
  name?: string
}

export interface ProcessData {
  total: number
  retList: ProcessResult[]
}

/** 查询结果 */
export interface ProcessResult {
  /** ID */
  id: number
  /** 名称 */
  name: string
  procInfo: ProcessNode[]
  nodeData: any[]
  edgeData: any[]
  status: number
  /** 修改时间 */
  updatedTime: string
  /** 创建时间 */
  createdTime: string
}

/** 表单对象 */
export interface ProcessForm {
  /** ID */
  id?: number
  /** 名称 */
  name?: string
  nodeData?: any[]
  edgeData?: any[]
  status?: number
}

export interface ProcessNodeForm {
  /** 名称 */
  name: string
  role: string
  roleName: string
  approvalType: number
  approvalEdit: number
}

export interface ProcessOptResult {
  id: number
}

export interface ProcessNode {
  name: string
  type: string
  deptName: string
  approvalType: number
  approvalEdit: number
  approver: string
  status: number
  approvalInfo: ApprovalInfo[]
}

export interface ApprovalInfo {
  approver: string
  approverName: string
  role: string
  roleName: string
  status: number
}

File: app/internal/store/model/process.go
package model

import "gorm.io/datatypes"

func NewProcess() *Process {
	return &Process{Model: &Model{}}
}

type Process struct {
	ID       int64          `json:"id" gorm:"column:id"`
	Name     string         `json:"name" gorm:"column:name" `
	ProcInfo datatypes.JSON `json:"ProcInfo" gorm:"column:proc_info" type:"json"`
	NodeData datatypes.JSON `json:"nodeData" gorm:"column:node_data" type:"json"`
	EdgeData datatypes.JSON `json:"edgeData" gorm:"column:edge_data" type:"json"`
	Status   int8           `json:"status" gorm:"column:status"`
	*Model
}

func (m *Process) TableName() string {
	return "process"
}

func (m *Process) IDField() string {
	return "id"
}
func (m *Process) NameField() string {
	return "name"
}
func (m *Process) ProcInfoField() string {
	return "proc_info"
}
func (m *Process) NodeDataField() string {
	return "node_data"
}
func (m *Process) EdgeDataField() string {
	return "edge_data"
}
func (m *Process) StatusField() string {
	return "status"
}

type PorcessNode struct {
	Name     string `json:"name"`
	DeptName string `json:"deptName"`
	// 节点类型 procStart procEnd procApproval
	Type string `json:"type"`
	// 审批人，默认时候为空，只有在指定审批人的时候在生效
	Approver string `json:"approver"`
	// 0 一个子节点通过就全通过，1 所有的通过才通过
	ApprovalType int8 `json:"approvalType"`
	// 0 审批时不可编辑，1 审批时编辑
	ApprovalEdit int8 `json:"approvalEdit"`
	// 节点的审批信息
	ApprovalInfo []*ApprovalInfo `json:"approvalInfo"`
	// 总的状态
	Status int8 `json:"status"`
}
type ApprovalInfo struct {
	Approver     string `json:"approver"`
	ApproverName string `json:"approverName"`
	Role         string `json:"role"`
	RoleName     string `json:"roleName"`
	// 审批节点状态
	Status int8 `json:"status"`
}
